File: Tutorial3-3LiveData-DataBinding-Transformations/src/main/java/com/example/tutorial3livedataanddatabinding2/MainActivity.java
package com.example.tutorial3livedataanddatabinding2;

import androidx.lifecycle.ViewModelProviders;
import androidx.databinding.DataBindingUtil;
import android.os.Bundle;
import androidx.appcompat.app.AppCompatActivity;

import com.example.tutorial3livedataanddatabinding2.databinding.ActivityMainBinding;
import com.example.tutorial3livedataanddatabinding2.viewmodel.MyViewModel;

public class MainActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);


        MyViewModel myViewModel = ViewModelProviders.of(MainActivity.this).get(MyViewModel.class);
        myViewModel.setUpLiveData();

        ActivityMainBinding binding = DataBindingUtil.setContentView(this, R.layout.activity_main);
        binding.setViewModel(myViewModel);
        // Required for binding with LiveData
        binding.setLifecycleOwner(this);

        /*
           myViewModel.getUsersLiveData().observe(this, new Observer<List<User>>() {
            @Override
            public void onChanged(@Nullable List<User> users) {
                binding.textView.setText(myViewModel.getUserRecords(users));
            }
        });
        */


    }
}

File: Tutorial6-2-Coroutine-LiveData-StateFlow-SharedFlow/src/main/java/com/omeriyioz/tutorial6_2_coroutine_livedata_stateflow_sharedflow/MainActivity.kt
package com.omeriyioz.tutorial6_2_coroutine_livedata_stateflow_sharedflow

import android.os.Bundle
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.lifecycleScope
import com.omeriyioz.tutorial6_2_coroutine_livedata_stateflow_sharedflow.databinding.ActivityMainBinding
import kotlinx.coroutines.flow.collect

class MainActivity : AppCompatActivity() {
    private val counterViewModel: MainViewModel by viewModels()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val binding: ActivityMainBinding = DataBindingUtil.setContentView(this, R.layout.activity_main)

        // 🔥 ‍LifeCycleOwner should be set for LiveData changes to be propagated to UI for this binding
        binding.lifecycleOwner = this

        binding.viewModel = counterViewModel


        // 😎 Setting StateFlow value in IO thread is safe.
        binding.buttonForStateFlow.setOnClickListener {
            counterViewModel.actionForStateFlow()
        }

        // Observe StateFlow
        lifecycleScope.launchWhenResumed {
            counterViewModel.weatherStateFlow.collect {
                binding.textViewStateFlow.text = it
            }
        }


        // 😎 Setting SharedFlow value in IO thread is safe.
        binding.buttonForSharedFlow.setOnClickListener {
            counterViewModel.actionForSharedFlow()
        }

        // Observe SharedFlow
        lifecycleScope.launchWhenResumed {
            counterViewModel.weatherSharedFlow.collect {
                binding.textViewSharedFlow.text = it
            }
        }

    }
}